### PR TITLE
Ambient sphinx geometry per mixnet

### PR DESCRIPTION
**Pull request by david415**
tl;dr, Add novel mixnet hybrid post quantum Sphinx. Improve code quality and improve architecture design. Make Sphinx API more flexible.

This feature set gets rid of our global default Sphinx Geometry and replaces it with the notion that every network actor has its own configuration with the correct Sphinx Geometry. We improve the Sphinx API and add a new "package" called core/sphinx/geo where we define the Geometry type AND constructors for creating the Geometry type for use with Sphinx. The Geometry type in turn knows how to create a Sphinx type, a Sphinx type is used for composing and unwrapping Sphinx packets. 

In our new mix network architecture, the Sphinx Geometry is determined by the dirauths, and given to other mix ops and clients out-of-band, just like the dirauth keys and network parameters are already distributed. Therefore everyone's config files have the identical Sphinx Geometry as every other participant in the mix network.

The PKI doc now contains a Sphinx Geometry hash value which not only enforces that all the dirauths agree on the geometry, but also allows servers and clients to fail fast if they somehow have incorrect Sphinx Geometry.

This allows for different Sphinx Geometry for different mixnets, as determined by the dirauths.

This feature set is the final architectural design change necessary for our post quantum cryptography upgrade. It allows for an upgrade to a fully post quantum mixnet which makes use of a hybrid post quantum Sphinx.

Our implementation of Sphinx is special in that it can use *any* NIKE (non-interactive key exchange) and *any* KEM (key encapsulation mechanism). Using a KEM with Sphinx implies additional Sphinx header overhead to store the KEM public key for each hop inside the header's routing info section. You are permitted to argue that this isn't really "Sphinx" but to me it is if it's called "Sphinx" in the code.

As of this commit, the NIKEs and KEMs we have available for use in Sphinx are the following:

```
var allSchemes = [...]nike.Scheme{
	csidh.CSIDHScheme,
	ecdh.NewEcdhNike(rand.Reader),
	hybrid.NOBS_CSIDHX25519,
	hybrid.CTIDHX25519,
}
```

and the KEMs from cloudflare's circl golang library:

```
var allSchemes = [...]kem.Scheme{
	hpke.KEM_P256_HKDF_SHA256.Scheme(),
	hpke.KEM_P384_HKDF_SHA384.Scheme(),
	hpke.KEM_P521_HKDF_SHA512.Scheme(),
	hpke.KEM_X25519_HKDF_SHA256.Scheme(),
	hpke.KEM_X448_HKDF_SHA512.Scheme(),
	frodo640shake.Scheme(),
	kyber512.Scheme(),
	kyber768.Scheme(),
	kyber1024.Scheme(),
	hybrid.Kyber512X25519(),
	hybrid.Kyber768X25519(),
	hybrid.Kyber768X448(),
	hybrid.Kyber1024X448(),
}
```

And let me remind you that any NIKE can be made into a KEM. And any two KEMs can be made into a hybrid KEM. Any two NIKEs can be made into a hybrid NIKE. And of course adding new KEMs and NIKEs is easy because they just have to adhere to a small set of simple interfaces: Scheme, PrivateKey and PublicKey... so named by cloudflare's circl library, I have continued the tradition in the Katzenpost NIKE.

### CODE REVIEW

**Review by leif — CHANGES_REQUESTED** [review]

File: authority/,1 [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+11:04:09.086 DEBU state2: Checking signature from 8642b6de0e42560874e08a5e20ae0a65a7fbfee405915d662a72a0d40990af6e on our certificates
+11:04:09.584 DEBU state2: Checking signature from 47fd47e22ce0d49a6c696dedd5cd33d3a4762073f7e680be8a2033f78ee08237 on our certificates
+11:04:10.647 NOTI state2: Consensus signed by 47fd47e22ce0d49a6c696dedd5cd33d3a4762073f7e680be8a2033f78ee08237
+11:04:10.647 NOTI state2: Consensus signed by 652036c88c6bcc27fd3020ce8aa5a78e4a4c59ead795a4687fe848b3102a04d9
+11:04:10.647 NOTI state2: Consensus signed by 8642b6de0e42560874e08a5e20ae0a65a7fbfee405915d662a72a0d40990af6e
+11:04:10.647 NOTI state2: Consensus made for epoch 150965 with 3/3 signatures
+--- PASS: TestVote (67.70s)
+PASS
+ok  	github.com/katzenpost/katzenpost/authority/voting/server	(cached)
+?   	github.com/katzenpost/katzenpost/authority/voting/server/config	[no test files]
```
[the remark this answers is not in the document]
Comment (david415) [reply authority/,1]: yeup. rm'ed

File: genconfig/main.go [hunk 2]
```diff
@@ -450,8 +450,8 @@ func main() {
 
 	nrHops := *nrLayers + 2
 
-	s.sphinxGeometry = geo.GeometryFromUserForwardPayloadLength(
-		ecdh.NewEcdhNike(rand.Reader),
+	s.sphinxGeometry = geo.KEMGeometryFromUserForwardPayloadLength(
+		hybrid.Kyber768X25519(),
```
[the remark this answers is not in the document]
Comment (david415) [reply genconfig/main.go]: yeup. added kem and nike CLI flags.

**Comment by leif** [comment]
note: this includes everything in #244 and #245.

**Comment by leif** [comment]
note: this branch briefly had some commits which depend on (and are needed by) #250 - i force pushed here to remove them so that this can land without that.

File: core/sphinx/geo/geo.go [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+
+	"github.com/katzenpost/katzenpost/core/crypto/nike"
+	"github.com/katzenpost/katzenpost/core/crypto/nike/schemes"
+	"github.com/katzenpost/katzenpost/core/sphinx/internal/crypto"
+)
+
+const (
+	NodeIDLength      = 32
+	RecipientIDLength = 32
+	SURBIDLength      = 16
```
Comment (leif) [inline core/sphinx/geo/geo.go:24]: these 3 values are still also defined (with the same values as here) in `core/sphinx/constants.go`; presumably `constants.go` should be removed in favor of this?

**Review by leif — CHANGES_REQUESTED** [review]

**Comment by leif** [comment]
@david415 i edited your opening comment here a bit, since it will become the commit message of the merge commit. Please review my changes to it.

**Review by leif — APPROVED** [review]
I think this is ready to land.

**Comment by leif** [comment]
closes #246

File: core/sphinx/internal/crypto/crypto.go [hunk 4]
```diff
@@ -166,8 +169,8 @@ func (k *PacketKeys) Reset() {
 // HOWEVER if privateKeySize is set to zero (0) then the BlindingFactor
 // will not be generated by the KDF. The reason we do this is because
 // KEMSphinx doesn't need the BlindingFactor.
```
Comment (leif) [inline core/sphinx/internal/crypto/crypto.go:171]: this comment isn't accurate anymore, it is now based on if `scheme != nil`

**Review by leif — CHANGES_REQUESTED** [review]

File: core/sphinx/internal/crypto/crypto.go (hunk 4 above)
Comment (david415) [reply core/sphinx/internal/crypto/crypto.go:171]: resolved

**Comment by david415** [comment]
> note: this branch briefly had some commits which depend on (and are needed by) #250 - i force pushed here to remove them so that this can land without that. after this lands, [e8f5eb7](https://github.com/katzenpost/katzenpost/commit/e8f5eb76ae0b18292979903017f84a0026a412a4) can be pushed to the branch at #250 (but if it is pushed there before this lands, then this PR's history will continue appear in the PR diff there due to github shortcomings...)

Note my edits to the Pull Request's comment where I briefly outline Sphinx API design changes and the architecture design changes. Not sure what else to write... unless you want to add something?

File: core/sphinx/geo/geo.go [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+	kemschemes "github.com/cloudflare/circl/kem/schemes"
+
+	"github.com/katzenpost/katzenpost/core/crypto/nike"
+	"github.com/katzenpost/katzenpost/core/crypto/nike/schemes"
+	"github.com/katzenpost/katzenpost/core/sphinx/constants"
+	"github.com/katzenpost/katzenpost/core/sphinx/internal/crypto"
+)
+
+const (
+	surbReplyLength = 1 + constants.SURBIDLength
```
Comment (threebithacker) [inline core/sphinx/geo/geo.go:23]: I'm assuming the `1 +` in these constants are 1-byte type tags? could we put a comment explaining what they are? (and perhaps the name of the function that decodes them?)

File: core/sphinx/kemsphinx_test.go [hunk 6]
```diff
@@ -40,30 +40,57 @@ func TestKEMSphinxSimple(t *testing.T) {
 	t.Parallel()
 	mykem := hybrid.Kyber768X25519()
 	withSURB := false
-	geo := KEMGeometryFromUserForwardPayloadLength(mykem, 512, withSURB, 5)
-	sphinx := NewKEMSphinx(mykem, geo)
+	g := geo.KEMGeometryFromUserForwardPayloadLength(mykem, 512, withSURB, 5)
```
Comment (threebithacker) [inline core/sphinx/kemsphinx_test.go:43]: here is a case where we aren't using (or testing) `2000` for example?

File: core/sphinx/kemsphinx_test.go [hunk 7]
```diff
@@ -138,13 +165,13 @@ func testForwardKEMSphinx(t *testing.T, mykem kem.Scheme, sphinx *Sphinx, testPa
...
-		pkt, err := sphinx.NewKEMPacket(rand.Reader, path, payload)
+		pkt, err := sphinx.NewPacket(rand.Reader, path, payload)
 		require.NoError(err, "NewKEMPacket failed")
-		require.Len(pkt, sphinx.Geometry().HeaderLength+sphinx.Geometry().PayloadTagLength+len(payload), "Packet Length")
+		require.Equal(len(pkt), sphinx.Geometry().HeaderLength+sphinx.Geometry().PayloadTagLength+len(payload), "Packet Length")
 
 		// Unwrap the packet, validating the output.
 		for i := range nodes {
-			b, _, cmds, err := sphinx.KEMUnwrap(nodes[i].privateKey, pkt)
+			b, _, cmds, err := sphinx.unwrapKem(nodes[i].privateKey, pkt)
```
Comment (threebithacker) [inline core/sphinx/kemsphinx_test.go:174]: we're not testing the `replayTag` here. can we do something useful with that? like maybe test that they are different in the various layers or something?

File: core/sphinx/sphinx_benchmark_test.go [hunk 8]
```diff
@@ -64,7 +60,7 @@ func benchNewNode(mynike nike.Scheme) *nodeParams {
 	}
 	n.publicKey, n.privateKey, err = mynike.GenerateKeyPair()
 	if err != nil {
-		panic(err)
+		panic("wtf")
```
Comment (threebithacker) [inline core/sphinx/sphinx_benchmark_test.go:63]: as fitting as `"wtf"` might be in this situation, the `err` is probably more useful?

File: core/sphinx/surb.go [hunk 9]
```diff
@@ -29,9 +29,7 @@ const (
 	sprpKeyMaterialLength = crypto.SPRPKeyLength + crypto.SPRPIVLength
 )
 
-// NewSURB creates a new SURB with the provided path using the provided entropy
-// source, and returns the SURB and decrypion keys.
-func (s *Sphinx) NewSURB(r io.Reader, path []*PathHop) ([]byte, []byte, error) {
+func (s *Sphinx) newNikeSURB(r io.Reader, path []*PathHop) ([]byte, []byte, error) {
```
Comment (threebithacker) [inline core/sphinx/surb.go:32]: we should keep this docstring too

**Review by threebithacker — CHANGES_REQUESTED** [review]

File: authority/voting/server/state_test.go [hunk 10]
```diff
@@ -574,11 +587,15 @@ func genMixConfig(name string, pki *sConfig.PKI, port uint16) (*identityKey, *sC
...
 
 // create epoch keys
-func genMixKeys(votingEpoch uint64) map[uint64]*ecdh.PublicKey {
-	mixKeys := make(map[uint64]*ecdh.PublicKey)
+func genMixKeys(votingEpoch uint64) map[uint64][]byte {
+	mixKeys := make(map[uint64][]byte)
 	for i := votingEpoch; i < votingEpoch+2; i++ {
-		idKey, _ := ecdh.NewKeypair(rand.Reader)
-		mixKeys[i] = idKey.PublicKey()
+		_, pubkey, err := ecdh.EcdhScheme.GenerateKeyPairFromEntropy(rand.Reader)
```
[the remark this answers is not in the document]
Comment (david415) [reply authority/voting/server/state_test.go:593]: MixKeys must be bytes so that it's generic. For this specific test it doesn't matter the contents of MixKeys, they don't need to be keys but they are signed in the test.

File: authority/voting/client/client_test.go [hunk 11]
```diff
@@ -60,14 +60,14 @@ func generateRandomTopology(nodes []*descriptor, layers int) [][]*pki.MixDescrip
...
+func generateMixKeys(epoch uint64) (map[uint64][]byte, error) {
+	m := make(map[uint64][]byte)
 	for i := epoch; i < epoch+3; i++ {
-		privatekey, err := ecdh.NewKeypair(rand.Reader)
+		_, privatekey, err := ecdh.EcdhScheme.GenerateKeyPairFromEntropy(rand.Reader)
 		if err != nil {
 			return nil, err
 		}
-		m[uint64(i)] = privatekey.PublicKey()
+		m[uint64(i)] = privatekey.Public().Bytes()
```
[the remark this answers is not in the document]
Comment (david415) [reply authority/voting/client/client_test.go:70]: I didn't think it mattered but i've fixed this now.

File: server/internal/scheduler/queue_bolt.go [hunk 12]
```diff
@@ -87,6 +86,8 @@ func packetToBoltBkt(parentBkt *bolt.Bucket, pkt *packet.Packet, prio time.Durat
 		bkt.Put([]byte(boltPacketPayloadKey), payloadBuf)
 	}
 
+	boltPacketCommandsSize := pkt.Geometry.NextNodeHopLength
```
[the remark this answers is not in the document]
Comment (david415) [reply server/internal/scheduler/queue_bolt.go:89]: fits perfectly:
```
		NextNodeHopLength:           1 + constants.NodeIDLength + crypto.MACLength,
```

```
// NextNodeHop is a de-serialized Sphinx next_node command.
type NextNodeHop struct {
	ID  [constants.NodeIDLength]byte
	MAC [crypto.MACLength]byte
}

```

File: memspool/server/spool_test.go [hunk 13]
```diff
@@ -110,8 +112,13 @@ func TestPersistence(t *testing.T) {
 	spoolID, err := spoolMap.CreateSpool(privKey.PublicKey(), signature)
 	assert.NoError(err)
 	messages := make([][]byte, 1)
+
+	mynike := ecdh.NewEcdhNike(rand.Reader)
+	nrHops := 5
+	geo := geo.GeometryFromUserForwardPayloadLength(mynike, 2000, true, nrHops)
```
[the remark this answers is not in the document]
Comment (david415) [reply memspool/server/spool_test.go:118]: Ooookay I've change the payload size in the ecdh sphinx test.

File: docker/Makefile [hunk 14]
```diff
@@ -33,6 +33,14 @@ lDMax=3000
 lM=0.2
 lMMax=100
 
+UserForwardPayloadLength=2000
```
[the remark this answers is not in the document]
Comment (david415) [reply docker/Makefile:36]: you just now said you don't want all the tests to be the same size. I've changed the payload size of the ecdh sphinx test to be 50000 bytes.

The payload size is available via the geometry object. We don't want to make variables for a default sphinx geometry since the whole point of the ticket is to make it so that the Sphinx geometry is controlled by the dirauths and is present in everyone's configuration. But some clients might want their own variable verify against that geometry object. In that specific context.

File: core/sphinx/kemsphinx_test.go (hunk 7 above)
Comment (david415) [reply core/sphinx/kemsphinx_test.go:174]: Why would we do that? Would we be simulating the behavior of the mix server? Would we be checking to make sure Sphinx is implemented correctly? If we look at how it's implemented we can decide a lot of stuff without writing tests.

File: core/sphinx/new_sphinx.go [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+func (s *Sphinx) Unwrap(privKey interface{}, pkt []byte) ([]byte, []byte, []commands.RoutingCommand, error) {
+	if s.nike == nil {
+		return s.unwrapKem(privKey.(kem.PrivateKey), pkt)
+	}
+
+	if s.kem == nil {
+		return s.unwrapNike(privKey.(nike.PrivateKey), pkt)
+	}
+
+	panic("Unwrap error, KEM and NIKE both cannot be onil")
```
[the remark this answers is not in the document]
Comment (david415) [reply core/sphinx/new_sphinx.go:53]: fixed

File: core/sphinx/sphinx.go [hunk 16]
```diff
@@ -440,9 +254,7 @@ func (s *Sphinx) createHeader(r io.Reader, path []*PathHop) ([]byte, []*sprpKey,
 	return hdr, sprpKeys, nil
 }
 
-// NewPacket creates a forward Sphinx packet with the provided path and
-// payload, using the provided entropy source.
-func (s *Sphinx) NewPacket(r io.Reader, path []*PathHop, payload []byte) ([]byte, error) {
+func (s *Sphinx) newNikePacket(r io.Reader, path []*PathHop, payload []byte) ([]byte, error) {
```
[the remark this answers is not in the document]
Comment (david415) [reply core/sphinx/sphinx.go:257]: private methods don't get docstrings. we should have a linter for this.

File: core/sphinx/sphinx_benchmark_test.go (hunk 8 above)
Comment (david415) [reply core/sphinx/sphinx_benchmark_test.go:63]: fixed

File: core/sphinx/surb.go (hunk 9 above)
Comment (david415) [reply core/sphinx/surb.go:32]: We do keep it in the public method.

Comment (david415) [reply core/sphinx/surb.go:32]: Please do and see new_sphinx.go where we have:
```
// NewSURB creates a new SURB with the provided path using the provided entropy
// source, and returns the SURB and decrypion keys.
func (s *Sphinx) NewSURB(r io.Reader, path []*PathHop) ([]byte, []byte, error) {

```

File: core/sphinx/geo/geo.go (hunk 5 above)
Comment (david415) [reply core/sphinx/geo/geo.go:23]: okay i've added another sphinx constant for that

File: core/sphinx/geo/geo.go [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+		HeaderLength:                f.headerLength(),
+		PacketLength:                f.packetLength(),
+		SURBLength:                  f.surbLength(),
+		UserForwardPayloadLength:    userForwardPayloadLength,
+		ForwardPayloadLength:        forwardPayloadLength,
+		PayloadTagLength:            payloadTagLength,
+		SphinxPlaintextHeaderLength: sphinxPlaintextHeaderLength,
+		RoutingInfoLength:           f.routingInfoLength(),
+		PerHopRoutingInfoLength:     f.perHopRoutingInfoLength(),
+		NextNodeHopLength:           1 + constants.NodeIDLength + crypto.MACLength,
```
[the remark this answers is not in the document]
Comment (david415) [reply core/sphinx/geo/geo.go]: also fixed here as well

File: core/sphinx/kemsphinx.go [hunk 18]
```diff
@@ -261,7 +264,7 @@ func (s *Sphinx) KEMUnwrap(privKey kem.PrivateKey, pkt []byte) ([]byte, []byte,
 	// Derive the various keys required for packet processing.
 	// note we set the private key size to zero because we do not
 	// derive blinding factors for KEMSphinx!
-	keys := crypto.KDF(sharedSecret, 0, nil)
+	keys := crypto.KDF(sharedSecret, nil)
```
[the remark this answers is not in the document]
Comment (david415) [reply core/sphinx/kemsphinx.go:267]: fixed

File: core/sphinx/kemsphinx.go [hunk 19]
```diff
@@ -133,7 +139,7 @@ func (s *Sphinx) createKEMHeader(r io.Reader, path []*PathHop) ([]byte, []*sprpK
 		// set privateKey size to zero 0 since
```
[the remark this answers is not in the document]
Comment (david415) [reply core/sphinx/kemsphinx.go:139]: fixed

File: core/sphinx/kemsphinx_test.go (hunk 6 above)
Comment (david415) [reply core/sphinx/kemsphinx_test.go:43]: lol yes of course it's works with ANY payload size. And here's another example of that. If you need more examples I can modify more tests.

File: core/pki/descriptor.go [hunk 20]
```diff
@@ -62,7 +67,7 @@ type MixDescriptor struct {
 	LinkKey wire.PublicKey
 
 	// MixKeys is a map of epochs to Sphinx keys.
-	MixKeys map[uint64]*ecdh.PublicKey
+	MixKeys map[uint64][]byte
```
[the remark this answers is not in the document]
Comment (david415) [reply core/pki/descriptor.go:70]: Raw bytes is the only way to do this.

File: core/sphinx/commands/commands.go [hunk 21]
```diff
@@ -145,12 +135,16 @@ func (cmd *Recipient) ToBytes(b []byte) []byte {
...
+
+	// recipientLength is the length of a Recipient command in bytes.
+	recipientLength := 1 + constants.RecipientIDLength
+
+	if len(b) < recipientLength-1 {
 		err = errInvalidCommand
 		return
 	}
-	rest = b[RecipientLength-1:]
+	rest = b[recipientLength-1:]
```
[the remark this answers is not in the document]
Comment (david415) [reply core/sphinx/commands/commands.go:148]: fixed

File: core/sphinx/commands/commands.go [hunk 22]
```diff
@@ -171,12 +165,16 @@ func (cmd *SURBReply) ToBytes(b []byte) []byte {
...
+
+	// surbReplyLength is the length of a SURBReply command in bytes.
+	surbReplyLength := 1 + constants.SURBIDLength
+
+	if len(b) < surbReplyLength-1 {
 		err = errInvalidCommand
 		return
 	}
-	rest = b[SURBReplyLength-1:]
+	rest = b[surbReplyLength-1:]
```
[the remark this answers is not in the document]
Comment (david415) [reply core/sphinx/commands/commands.go:178]: fixed

File: core/sphinx/commands/commands.go [hunk 23]
```diff
@@ -200,11 +198,14 @@ func (cmd *NodeDelay) ToBytes(b []byte) []byte {
...
-	if len(b) < NodeDelayLength-1 {
+	// nodeDelayLength is the length of a NodeDelay command in bytes.
+	nodeDelayLength := 1 + 4
+
+	if len(b) < nodeDelayLength-1 {
 		err = errInvalidCommand
 		return
 	}
-	rest = b[NodeDelayLength-1:]
+	rest = b[nodeDelayLength-1:]
```
[the remark this answers is not in the document]
Comment (david415) [reply core/sphinx/commands/commands.go:209]: fixed

File: core/sphinx/geo/geo.go (hunk 5 above)
Comment (leif) [reply core/sphinx/geo/geo.go:23]: you added a new constant but aren't using here on line 23 yet

File: core/sphinx/commands/commands.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (leif) [reply core/sphinx/commands/commands.go:170]: there is some github bug where it still isn't marking this diff as "outdated" but this is actually fixed now.

File: core/sphinx/commands/commands.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (leif) [reply core/sphinx/commands/commands.go:144]: there is some github bug where it still isn't marking this diff as "outdated" but this is actually fixed now.
